### ISSUE TITLE
Issues with multiple concurrent users

### ISSUE THREAD

**Issue by mjohnson767** [artilleryio/artillery#2147]
Hello Hassy, I'm very interested in the Artillery | Playwright testing but experiencing the same issue as # 1711.
Build: artillery/2.0.0-37 win32-x64 node-v18.14.0
My goal is run tests with the VUser concurrent limit with a ramp-up (to the limit) and duration for the test. This is pretty much standard for all load testing tools. For some quick testing I've set a low amount of Vusers to test it out. With the following setup the actual concurrent rate is only 1 VUser that just repeats until the duration is hit.

config:
target: "https://mytesturl.com/"
phases:
- duration: 60
rampTo: 3
name: Ramp up load
maxVusers: 3
payload:
- path: ./logins.csv
fields:
- "userid"
- "recordid"
order: sequence

If I take out the maxUsers I'll see several browsers open with only one of the actually going through the workflow.

Attempting to run with the WORKERS setting fails with this version. WORKERS=3 artillery run script.yaml
WORKERS=3 : The term 'WORKERS=3' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included,
verify that the path is correct and try again.

Do you have any ideas?
I've ran tests with Locust/Selenium locally and I know I can 20-30 browsers on my laptop. Running Artillery/Playwright should easily out perform locust/selenium. With a single load generator we were able to run 175 headless browsers. I'm hoping to surpass that with Artillery/Playwright.

### Discussed in https://github.com/artilleryio/artillery/discussions/1711

<div type='discussions-op-text'>

<sup>Originally posted by **JulieMatic** December  5, 2022</sup>
I am currently working on some tests that are using artillery-engine-playwright, and I'm running into a couple of different issues depending on the combination of versions I use. 

### Only the "active" Chrome window is interacted with
```
phases:
            - name: 10user
              duration: 300
              arrivalRate: 1
              arrivalCount: 1000
              maxVusers: 10
```
Using the above phase, artillery version 1.7.9, and artillery-engine-version 0.2.0, when I run `artillery run --environment 10_concurrent_users ./dist/tests/Performance/artillery/artillery-10-users-prod-intro.yaml`, over the course of a few seconds, I see 10 Chrome windows pop up, but only the "active" window gets interacted with. For example, the first thing my test does is enter an email address, password, and click "Sign in." 9 of the windows do nothing, and the "active" window will input multiple email addresses into the email field. For all 10 windows, the tests get stuck at the login page because either no credentials are entered, or invalid credentials are provided. 

_Note_: So that I can see what's happening more easily, I'm running with `headless: false`, but one of my coworkers seems to be running into a similar issue where the tests get stuck at the login page and he is setting `headless: true`

### Phase configuration not behaving the same as before
```
phases:
            - name: 10user
              duration: 300
              arrivalRate: 1
              arrivalCount: 1000
              maxVusers: 10
```
Using the above phase (same as the previous phase), artillery version 2.0.0-27, and artillery-engine-version 0.2.1, when I run the same `artillery run` command, I only see 1 user running at a time, despite `maxVusers` being set to 10. 

If I instead configure my phase as follows:
```
phases:
            - name: 10user
              duration: 300
              arrivalRate: 3
              maxVusers: 10
```
When I run the same `artillery run` command, I initially see 2 Chrome windows pop up, the tests for those 2 windows complete, and then 3 more windows pop up. Then, throughout the duration of the artillery test, the maximum number of concurrent users I see at any given time is 3. With this phase, I'm seeing the issue where the test is only interacting with the active issue go away, and the playwright scripts can finish as expected. 

Using a similar phase configuration, but changing `arrivalRate` to 10, when I run the same `artillery run` command, I get something like 4 Chrome windows popping up, the tests for those windows completing, and then a couple more Chrome windows being added as tests finish, until I reach the maximum of 10. 

This is leading to me to believe that something changed with the phase configuration options. My understanding of the phase configuration options (based on [this documentation](https://www.artillery.io/docs/guides/guides/test-script-reference)) is that:

- `arrivalRate` is used to specify how many virtual users to add each second, so if I have `duration` set to 300 and `arrivalRate` set to 3, I should expect to see 3 virtual users added every second for 5 minutes
- `arrivalCount` is used to specify how many virtual users to add over the duration, so if I have `duration` set to 300 and `arrivalCount` set to 1000, I'd expect to see 1 virtual user about every 3 seconds 
- I haven't been able to figure out if `arrivalRate` and `arrivalCount` can/should be used together

But instead, it seems like `arrivalRate` is behaving like `maxVusers` and that there is a built-in ramp-up behavior. 

### Questions

1. Is there a known issue with artillery-engine-playwright where running artillery with multiple concurrent virtual users will result in only the active window being interacted with?
2. Was there a change in the phase configuration options? Without changing my YAML file between switching versions, I'm seeing a difference in behavior between artillery version 1.7.9 and artillery version 2.0.0-27, but I'm not sure which version is behaving correctly. 
3. Using the phases I provided in this post as examples, what is the expected behavior?
4. What is the ideal combination of versions for artillery and artillery-engine-playwright?
</div>

**Comment by bernardobridge** [comment]
Hi @mjohnson767 👋 ,

I believe the error you're getting
```shell
The term 'WORKERS=3' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included,
verify that the path is correct and try again.
```
is from setting an environment variable with Powershell the way you would in a Linux environment (https://stackoverflow.com/questions/1420719/powershell-setting-an-environment-variable-for-a-single-command-only).

That being said, the bug you mentioned should be fixed already. On MacOS, I had no issues running a test with a config similar to yours - all chrome windows ran the test. I will try to run an Artillery Playwright test in a Windows environment as soon as I get the chance, but in the meantime:
- Can you make sure you're using Artillery v2.0.0-37, without installing the Playwright engine separately (it comes bundled now);

**Comment by mjohnson767** [comment]
I reset the environment with only the Artillery v2.0.0-37 install without the additional Playwright engine install. 
This phase certainly doesn't work:
  phases:
       duration: 60
       rampTo: 3
       name: Ramp up load
       maxVusers: 3
Several Browser windows would launch and for a split seconds attempt to launch the website and then shut down. 

My scenario requires unique logins (using 3 ID's for testing) so for 3 user ID's and so I would cap the VUsers at 3. This phase worked somewhat:
  phases:
      # ** will create 3 vusers in 9 seconds
    -  name: fixed_arrivals
       duration: 120
       arrivalRate: 3    
       maxVusers: 3

The VUsers would run through the complete scenario, which was great to see but the concurrency was a little odd. There were long pauses between each completed scenario and 75% of the time I there would be 2 VUsers running simultaneously and the other 25% all 3 VUsers would be running concurrently. 

The next two items I need to setup is setting up Grafana  so I can monitor the test. Then wrap transactional timers around each request that does a call to the server so I can gather all the needed metrics. I see the example posted in the flows.js script but I haven't figured out how to implement it in my script yet. If you have some better examples that would be great. 

Here's my current working script:
async function secure(page,userContext) {
  const userid = userContext.vars.userid;
  const recordid = userContext.vars.recordid;
  const startedTime = Date.now();
  
  // landing page TRANACTION TIMER 01_Landing page
  await page.goto('https://internaltesturl.com/ALTLOGIN');
  // const difference = Date.now() - startedTime;
  await page.waitForTimeout(1000);
  await page.getByLabel(`MS ID (example: asmith1)`).fill( `${userid}` );
  await page.getByLabel('Password').fill(`${password}`);

  await page.waitForTimeout(1000);

  // login TRANACTION TIMER 02_SubmitLogin
  await page.getByRole('button', { name: 'Submit' }).click();
  //const difference = Date.now() - startedTime;
  //events.emit('histogram', 'Login', difference);

  await page.waitForTimeout(1000);

  await page.getByPlaceholder('Enter Request ID (minimum 5 numeric characters)').fill(`${recordid}`);

  await page.waitForTimeout(1000);

  // const startedTime = Date.now();
  // Search Record TRANACTION TIMER 03_SearchRecord
  await page.getByRole('button', { name: ' Go' }).click();
  //const difference = Date.now() - startedTime;
  //events.emit('histogram', 'Search', difference);
 
  await page.waitForTimeout(1000);

  await page.locator('css=button.uitk-s-avatar-button.login').click();
  await page.waitForTimeout(1000);

  // const startedTime = Date.now();
  // Logout TRANACTION TIMER 04_Logout
  await page.getByText('Logout').click();
  //const difference = Date.now() - startedTime;
  //events.emit('histogram', 'Search', difference);
}

**Comment by bernardobridge** [comment]
Hey @mjohnson767 ! I edited your post to include [block quotes]() for readability.

I originally had thought you were having the problem with `arrivalRate` not `rampTo` (given the original issue you referenced was with `arrivalRate`). We'll investigate the behaviour more with `rampTo` 👍 

---

But it sounds like for now you are unblocked by using `arrivalRate`. So I would suggest using that for now (you can anyway create "ladders" of multiple phases of increasing arrival rates, which would somewhat mimic a ramp phase).

**Comment by bernardobridge** [comment]
Regarding this question:

>The next two items I need to setup is setting up Grafana so I can monitor the test. Then wrap transactional timers around each request that does a call to the server so I can gather all the needed metrics. I see the example posted in the flows.js script but I haven't figured out how to implement it in my script yet. If you have some better examples that would be great.

The transactional timers (emitting custom metrics) in the example should work. If you want to do multiple ones, then you would need to name the custom metrics differently though. I think in that case creating a utility function for it might make the code better. I've taken a stab at it for you:

### 1. Create a utility function for wrapping the playwright code that constitutes a transaction:

```js
const withTransactionTimer = async (transactionName, events, userActions) => {
  const startedTime = Date.now();
  await userActions();
  const difference = Date.now() - startedTime;

  events.emit('histogram', transactionName, difference)
}
```

### 2. Wrap the different transactions in your playwright code with this, giving different names to each

```js
async function secure(page,userContext) {
  const userid = userContext.vars.userid;
  const recordid = userContext.vars.recordid;

  await withTransactionTimer("landing_page", events, async () => {
    await page.goto('https://internaltesturl.com/ALTLOGIN');
    await page.waitForTimeout(1000);
    await page.getByLabel(`MS ID (example: asmith1)`).fill( `${userid}` );
    await page.getByLabel('Password').fill(`${password}`);
    await page.waitForTimeout(1000);
  })

  await withTransactionTimer("submit_login", events, async () => {
    await page.getByRole('button', { name: 'Submit' }).click();
    await page.waitForTimeout(1000);
    await page.getByPlaceholder('Enter Request ID (minimum 5 numeric characters)').fill(`${recordid}`);
    await page.waitForTimeout(1000);
  })

  await withTransactionTimer("search_record", events, async () => {
    await page.getByRole('button', { name: ' Go' }).click();
    await page.waitForTimeout(1000);
    await page.locator('css=button.uitk-s-avatar-button.login').click();
    await page.waitForTimeout(1000);
  })

  await withTransactionTimer("logout", events, async () => {
    await page.getByText('Logout').click();
  })
}
```

I think this makes the code much more readable. It will emit custom metrics for each of these transactions, named as the first argument of the `withTransactionTimer` function.

**Comment by bernardobridge** [comment]
@mjohnson767 , just realised what was probably the reason the custom events were not working for you:

in your `secure` function, you forgot `events` in the function signature:

```js
async function secure(page, userContext, events//this was missing) {
...
}
```

You need to pass that in, as otherwise the event emitters won't be available to emit the custom metrics. Hopefully things work for you now :) (and you should be able to use my code above to make it a repeatable pattern).

**Comment by mjohnson767** [comment]
This is the information I was looking for, I'll get these ideas implemented and will let you know.

**Comment by mjohnson767** [comment]
Works perfect. I'll research adding assertions and validations. I was taking data from codegen. 
--------------------------------
Summary report @ 12:14:26(-0600)
--------------------------------

01_landing_page:
  min: ......................................................................... 3187
  max: ......................................................................... 6061
  median: ...................................................................... 3605.5
  p95: ......................................................................... 4147.4
  p99: ......................................................................... 4147.4
02_submit_login:
  min: ......................................................................... 2902
  max: ......................................................................... 5878
  median: ...................................................................... 3134.5
  p95: ......................................................................... 5711.5
  p99: ......................................................................... 5711.5
03_search_record:
  min: ......................................................................... 2061
  max: ......................................................................... 4687
  median: ...................................................................... 2101.1
  p95: ......................................................................... 3605.5
  p99: ......................................................................... 3605.5
04_logout:
  min: ......................................................................... 18
  max: ......................................................................... 77
  median: ...................................................................... 26.8
  p95: ......................................................................... 50.9
  p99: ......................................................................... 50.9
browser.http_requests: ......................................................... 242

**Comment by bernardobridge** [comment]
For anyone checking this thread, we've released in [Artillery v2.0.0-38](https://github.com/artilleryio/artillery/releases/tag/v2.0.0-38) a convenience helper similar to the one exemplified above, allowing you to wrap parts of your Playwright code with step() that emit Artillery custom metrics.

More information in: https://www.artillery.io/docs/reference/engines/playwright#teststep-argument
